mtp/mtp_db.cpp

```diff
@@ -1,57 +1,79 @@
+        {
+            return iter->second;
+        }
+        inline const Handle &getHandle(
+            std::map<Handle, std::map<std::filesystem::path, Handle>::iterator>::const_iterator iter)
+        {
+            return iter->first;
+        }
+        inline const std::filesystem::path &getFilename(
+            std::map<Handle, std::map<std::filesystem::path, Handle>::iterator>::const_iterator iter)
+        {
+            return iter->second->first;
+        }
+
+    } // namespace handle_to_filename
+
     std::optional<std::filesystem::path> FileDatabase::get_filename(Handle handle) const
     {
-        if (const auto result = entries.find(DatabaseEntry{handle}); result != entries.end()) {
-            return result->get_filename();
+        const auto handleToFilenameIter = handleToFilename.find(handle);
+        if (handleToFilenameIter != handleToFilename.end()) {
+            return handle_to_filename::getFilename(handleToFilenameIter);
         }
         return std::nullopt;
     }
     bool FileDatabase::remove(const Handle handle)
     {
-        return entries.erase(DatabaseEntry{handle}) != 0;
+        const auto handleToFilenameIter = handleToFilename.find(handle);
+        if (handleToFilenameIter != handleToFilename.end()) {
+            filenameToHandle.erase(handle_to_filename::getIter(handleToFilenameIter));
+            handleToFilename.erase(handleToFilenameIter);
+            return true;
+        }
+        return false;
     }
-    std::optional<Handle> FileDatabase::insert(const char *filename)
+    Handle FileDatabase::insert(const char *filename)
     {
         static Handle handle_idx = 1;
-        if (const auto result = entries.insert(DatabaseEntry{handle_idx++, filename}); result.second) {
-            return result.first->get_handle();
+
+        const auto entry = filenameToHandle.emplace(filename, handle_idx);
+        if (entry.second) {
+            handleToFilename.emplace(handle_idx, entry.first);
+            ++handle_idx;
         }
-        return std::nullopt;
+        return filename_to_handle::getHandle(entry.first);
     }
     bool FileDatabase::update(const Handle handle, const char *filename)
     {
-        if (const auto result = entries.find(DatabaseEntry{handle}); result != entries.end()) {
-            remove(result->get_handle());
-            entries.insert(DatabaseEntry{result->get_handle(), filename});
+        const auto handleToFilenameIter = handleToFilename.find(handle);
+        if (handleToFilenameIter != handleToFilename.end()) {
+            filenameToHandle.erase(handle_to_filename::getIter(handleToFilenameIter));
+            auto entry = filenameToHandle.emplace(filename, handle);
+            if (!entry.second) {
+                filename_to_handle::getHandle(entry.first) = handle_to_filename::getHandle(handleToFilenameIter);
+            }
+            handle_to_filename::getIter(handleToFilenameIter) = entry.first;
```

Comment:
In the previous implementation we had a call to `remove()`, why is none such here?

Reply:
In the previous implementation, function remove() was simply doing erase(), right now as it has two maps to keep the data we are doing erase in place as function remove() is doing more, and there is no need to remove all records and add them once more within update operation.

Reply:
Fine